stream.go

```diff
@@ -180,12 +180,30 @@ func NewClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, meth
 var emptyMethodConfig = serviceconfig.MethodConfig{}
 
 func newClientStream(ctx context.Context, desc *StreamDesc, cc *ClientConn, method string, opts ...CallOption) (_ ClientStream, err error) {
+	if channelz.IsOn() {
+		cc.incrCallsStarted()
+	}
+	defer func() {
+		if err != nil {
+			if channelz.IsOn() {
+				cc.incrCallsFailed()
+			}
+			// Invoke all the registered OnFinish call options explicitly. A
+			// non-nil error means that the stream wasn't created, and
+			// therefore these will be NOT be invoked as part of `cs.finish()`.
+			for _, o := range opts {
+				if o, ok := o.(OnFinishCallOption); ok {
+					o.OnFinish(err)
+				}
+			}
```

Comment:
Let's see if we can find a better way to unify these things?

Maybe we can have a shared endOfRPC function that is called by finish and from here?